feathr_project/feathr/_feature_registry.py

```diff
@@ -838,3 +839,142 @@ def _list_registered_entities_with_details(self, project_name: str = None, entit
             guid=guid_list)["entities"]
         return entity_res
+        
+    def get_features_from_registry(self, project_name: str) -> Tuple[List[FeatureAnchor], List[DerivedFeature]]:
+        """[Sync Features from registry to local workspace, given a project_name, will write project's features from registry to to user's local workspace]
+        Args:
+            project_name (str): project name.
+        """
+
+        entities = self._list_registered_entities_with_details(project_name=project_name, entity_type=TYPEDEF_FEATHR_PROJECT)
+        if not entities:
+            # if the result is empty
+            return (None, None)
+        project_entity = entities[0] # there's only one available
+        anchor_guid = [anchor_entity["guid"] for anchor_entity in project_entity["attributes"]["anchor_features"]]
+        derived_feature_guid = [derived_feature_entity["guid"] for derived_feature_entity in project_entity["attributes"]["derived_features"]]
+        
+        derived_feature_result = self.purview_client.get_entity(guid=derived_feature_guid)["entities"]
+        derived_feature_list = []
+        for derived_feature_entity in derived_feature_result:
+            key_from_entity=derived_feature_entity["attributes"]["tags"]
+            anchor_feature_guid = [e["guid"] for e in derived_feature_entity["attributes"]["input_anchor_features"]]
+            derived_feature_guid = [e["guid"] for e in derived_feature_entity["attributes"]["input_derived_features"]]
+            input_features_guid = self.search_input_anchor_features_bfs(derived_feature_guid)
+            all_input_features = list(itertools.chain.from_iterable([self._get_features_by_guid(x) for x in input_features_guid+anchor_feature_guid]))
+
+            derived_feature_list.append(DerivedFeature(name=derived_feature_entity["attributes"]["name"],
+                                feature_type=None,
+                                transform=None,
+                                key=key_from_entity,
+                                input_features= all_input_features,
+                                registry_tags=derived_feature_entity["attributes"]["tags"]))
+        anchor_result = self.purview_client.get_entity(guid=anchor_guid)["entities"]
+        anchor_list = []
+        for anchor_entity in anchor_result:
+            feature_guid = [e["guid"] for e in anchor_entity["attributes"]["features"]]
+            anchor_list.append(FeatureAnchor(name=anchor_entity["attributes"]["name"],
+                                source=self._get_source_by_guid(anchor_entity["attributes"]["source"]["guid"]),
+                                features=self._get_features_by_guid(feature_guid),
+                                registry_tags=anchor_entity["attributes"]["tags"]))
+
+        
+        return (anchor_list, derived_feature_list)
+
+    def search_input_anchor_features_bfs(self,derived_guids) ->List[str]:
```

Comment:
docs

Reply:
i am not sure why we need to mention BFS. technical details should be hidden.

Reply:
changed